eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs

```diff
@@ -122,8 +129,10 @@ instance
   type Event (UTXOS era) = UtxosEvent era
   transitionRules = [utxosTransition]
 
-newtype UtxosEvent era
-  = UpdateEvent (Event (Core.EraRule "PPUP" era))
+data UtxosEvent era
+  = AlonzoPpupToUtxosEvent (Event (Core.EraRule "PPUP" era))
+  | SuccessfulPlutusScriptsEvent [PlutusDebug]
+  | FailedPlutusScriptsEvent [PlutusDebug]
```

Comment:
I'd suggest using `NonEmpty` here as well
```suggestion
  | FailedPlutusScriptsEvent (NonEmpty PlutusDebug)
```

This will be useful for downstream consumers of this event.

Reply:
I actually tried to use a `NonEmpty` list here, but ran into an unfortunate lack of "parse, don't validate". I'd love to fix it, I'm not sure how easy it will be. Maybe a follow up PR?

The function `collectTwoPhaseScriptInputs` only returns Plutus scripts (together will the collected data that it needs to run them) in the happy path, but unfortunately the returns is `AlonzoScript.Script`, thus forgetting the fact that only Plutus scripts are present.  When `evalScripts` get the results from `collectTwoPhaseScriptInputs`, it has to do something with the case of Timelocks, so it creates the possibility of a 1-phase failure, namely the constructor `OnePhaseSF Text` of `ScriptFailure`.

This leads to having to filter the non-empty list of `ScriptFailure`s to just those relating to plutus. See the function `scriptFailuresToPlutusDebug` introduced in this PR. And this is why I could not easily use a non-empty list inside `FailedPlutusScriptsEvent`.

This is a silly situation, of course, and I do want to fix it.

Reply:
it's worth noting that timelocks (aka "1-phase scripts") are validated in Alonzo in:

https://github.com/input-output-hk/cardano-ledger/blob/70dbfc9f12283666e29f8492aff5da9b1c8c5a7f/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs#L361

This check uses the `validateScript` class method of `ValidateScript`. One of the first bits of tech debt that I would like to address post Babbage is an overhaul in the handling of 1-phase vs 2-phase scripts. The `validateScript` methods look generic, as though it runs all scripts, but does not.

Reply:
It is silly indeed. I don't think it make sense to have a possibility of emitting a `FailedPlutusScriptsEvent` without any failure. Correct me if I am wrong, but I think there is an easy solution to this, would this work?
https://github.com/input-output-hk/cardano-ledger/pull/2724

Reply:
I've just pushed another commit which addresses the root of the problem, namely the unneeded checking of timelocks inside collectTwoPhaseInputs.

Reply:
Nice!!!! :100: 
I think this is one of those examples where a stricter type encourages a better implementation

Reply:
yep, absolutely. I've gotten so much mileage out of [Parse, don’t validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).